src/session.rs

```diff
@@ -2152,23 +2152,31 @@ impl Session {
     }
 
     /// Get the color at the given view coordinate.
-    fn color_at(&self, v: ViewId, p: ViewCoords<u32>) -> Rgba8 {
+    fn color_at(&self, v: ViewId, p: ViewCoords<u32>) -> Option<Rgba8> {
         let resources = self.resources.lock();
         let snapshot = resources.get_snapshot(&v);
 
-        assert!(snapshot.len() % std::mem::size_of::<Bgra8>() == 0);
-
         let pixels = snapshot.pixels();
         let (head, pixels, tail) = unsafe { pixels.align_to::<Bgra8>() };
 
         assert!(head.is_empty());
         assert!(tail.is_empty());
 
-        let index =
-            ((snapshot.height() - p.y - 1) * snapshot.width() + p.x) as usize;
-        let bgra = pixels[index];
+        let y_offset = snapshot
+            .height()
+            .checked_sub(p.y)
+            .and_then(|x| x.checked_sub(1));
+        let index = y_offset.map(|y| (y * snapshot.width() + p.x) as usize);
 
-        Rgba8::new(bgra.r, bgra.g, bgra.b, bgra.a)
+        if let Some(index) = index {
+            if let Some(bgra) = pixels.get(index) {
+                Some(Rgba8::new(bgra.r, bgra.g, bgra.b, bgra.a))
+            } else {
+                None
+            }
+        } else {
+            None
+        }
```

Comment:
Can we perhaps make this a little tighter by using `and_then` to get rid of the `else` branches?

Reply:
Sure!